### DISCUSSION TITLE
SQL vs MongoDB

### DISCUSSION THREAD

**Discussion by ghost** [advplyr/audiobookshelf#1202] [Ideas]
Hi, 
I saw that the current DB is written in njodb. And it is mentioned in several issues that there is a transformation planned for SQLite.

Is it worthwhile to consider MongoDB? Since MongoDB is NOSQL, just like njodb, would that make the transformation easier?
Also, for SQL, PostgreSQL might perform better than SQLite. Would it be a worthwhile candidate?

Thank you.

**Comment by advplyr** [comment]
We need an embedded DB. There are a lot of embedded NoSQL db options and I fully implemented [acebase](https://github.com/appy-one/acebase) at one point but it was way too slow. We don't need all of those features and they are only making things slower.

If we could find an embedded NoSQL db it would be easier to migrate to but SQLLite is the most widely used and stable option.

I know MongoDB has the option to be embedded and I haven't done much research on that but I would guess it wouldn't perform as well as a db designed to be embedded.

**Reply by Gunsi1973** [reply]
so there is no "not embedded" db support planned?

**Reply by advplyr** [reply]
Possibly after we switch to a relational db. For a not embedded db the user will have to setup their own. The embedded db is for the vast majority of users who don't want to setup a separate db.
Allowing an external db is probably far down the road, I haven't really thought about it yet.

**Reply by ghost** [reply]
Since the rease is in source code format. I would imagine most users are in docker land. If not using docker, the rest users would be quite technical acquaintant in setting up server softwares from source. In either cases, it seems that external DB would most probably not be a problem. 

At the end of the day, this is just a design decision, trading in CPU and RAM resources for performance and scaling capabilities. 

We, as users, appreciate the huge amount of work you maintainers has done with this app. Whatever decisions you make, we will honor it and keep using your software.

**Reply by ghost** [reply]
Thank you for the awesome work.

**Reply by advplyr** [reply]
Most users now are technical and would be capable of running their own external db but it should not be a requirement. Once we have sqlite embedded db stable then we can discuss support external db's.
Eventually we will have a windows installer also so it won't always be all technical users. Hopefully a .NET developer joins to help us build that out.

**Reply by keaganhilliard** [reply]
We could potentially use an ORM like Prisma, which would allow support for all of these database backends without code changes. Start with getting the models right and then let Prisma handle the rest.
